modules/apps/frontend-data-set/frontend-data-set-impl/src/main/java/com/liferay/frontend/data/set/internal/renderer/DataSetRendererImpl.java

```diff
@@ -0,0 +1,70 @@
+/**
+ * SPDX-FileCopyrightText: (c) 2024 Liferay, Inc. https://liferay.com
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
+ */
+
+package com.liferay.frontend.data.set.internal.renderer;
+
+import com.liferay.frontend.data.set.renderer.DataSetRenderer;
+import com.liferay.frontend.data.set.renderer.ReactPropsProvider;
+import com.liferay.frontend.data.set.renderer.ReactPropsProviderRegistry;
+import com.liferay.portal.kernel.log.Log;
+import com.liferay.portal.kernel.log.LogFactoryUtil;
+import com.liferay.portal.template.react.renderer.ComponentDescriptor;
+import com.liferay.portal.template.react.renderer.ReactRenderer;
+
+import java.io.Writer;
+
+import java.util.Map;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Daniel Sanz
+ */
+@Component(service = DataSetRenderer.class)
+public class DataSetRendererImpl implements DataSetRenderer {
+
+	@Override
+	public void render(
+		boolean inline, String componentId, String propsTransformer,
+		String fdsName, HttpServletRequest httpServletRequest,
+		HttpServletResponse httpServletResponse, Map<String, Object> context,
+		Writer writer) {
+
+		try {
+			ReactPropsProvider reactPropsProvider =
+				_reactPropsProviderRegistry.getReactPropsProvider(
+					fdsName, httpServletRequest);
+
+			ComponentDescriptor componentDescriptor = new ComponentDescriptor(
+				"{FrontendDataSet} from frontend-data-set-web", componentId,
+				null, inline, propsTransformer);
+
+			_reactRenderer.renderReact(
+				componentDescriptor,
+				reactPropsProvider.getObjectEntryReactProps(
+					fdsName, httpServletRequest, httpServletResponse, context),
+				httpServletRequest, writer);
+		}
+		catch (Exception exception) {
+			if (_log.isWarnEnabled()) {
+				_log.warn("Unable to render data set " + fdsName, exception);
+			}
+		}
+	}
+
```

Comment:
Could we split off the whole renderer mechanism into a separate PR? First PR could just be `DataSet` interface, UI in DSM, and import to DSM persistence. We end with entry showing up in DSM, but is not applied in any of the existing usages. Second PR this mechanism. Parallel to second PR, we can extend first one with actions filters etc. The renderer mechanism I believe needs to implement all at once, for `Data Set` fragment to continue working.

Reply:
A high level sketch of possible PR distribution would be:

- DataSet interfaces + base implementation (creator, registries). We can do this at once or in parts (actions, policies, filters, fields, etc)
- A basic DSM UI, no designs yet, gated under FF
- Renderers, at once. We don't need to evolve the taglibs yet but I would consider working on it in parallel
- "Client Code" (expose customized data sets) for picklist, with FF. This one is straight forward
- "Client Code" (expose customized data sets) for products, with FF. This FF will take time as we need the Bulk actions to work

The last one is strictly not needed but we need some way to work on it. A possibility is to implement all those data sets in a sample module, kind of `frontend-data-set-sample-web` , then move to the corresponding modules later on with the help of the corresponding product teams.



- Basi